src/preprocessing/passes/symmetry_detect.cpp

```diff
@@ -313,147 +801,162 @@ void SymmetryDetect::collectChildren(Node node, vector<Node>& children)
+    for (unsigned index : nvi.second)
     {
-      subvars.resize(subvars.size()+(*part_it).d_subvar_to_vars.size());
+      Partition& p = partitions[index];
+      std::vector<Node>& svs = p.d_subvar_to_vars.begin()->second;
+      unsigned aindex = ntrie.add(index, svs);
+      if (aindex == index)
+      {
+        nvis.insert(index);
+      }
+      else if (theory::quantifiers::TermUtil::isNonAdditive(k))
+      {
+        Trace("sym-dt-debug") << "Drop duplicate child : " << index
+                              << std::endl;
+        Assert(active_indices.find(index) != active_indices.end());
+        active_indices.erase(index);
+      }
+      else
+      {
+        nvis.erase(aindex);
+      }
     }
+    std::vector<unsigned> check_indices;
+    check_indices.insert(check_indices.end(), nvis.begin(), nvis.end());
+    Trace("sym-dt-debug") << "Merge..." << std::endl;
+    // now, try to merge these partitions
+    mergePartitions(k, partitions, check_indices, active_indices);
   }
-
-  Trace("sym-dt") << "[sym-dt] Substitution variables for the variable "
-                  << target_var << ": " << subvars << endl;
-  addNode(target_var, subvars);
 }
 
-void SymmetryDetect::PartitionTrie::addNode(Node target_var,
-                                            vector<Node>& subvars)
+void SymmetryDetect::mergePartitions(
+    Kind k,
+    std::vector<Partition>& partitions,
+    const std::vector<unsigned>& indices,
+    std::unordered_set<unsigned>& active_indices)
 {
-  if (subvars.empty())
+  if (indices.size() <= 1)
   {
-    d_variables.push_back(target_var);
+    return;
   }
-  else
+  if (Trace.isOn("sym-dt-debug"))
   {
-    vector<Node> new_subs;
-    map<Node, PartitionTrie>::iterator children_it =
-        d_children.find(subvars[0]);
-
-    if (subvars.begin() + 1 < subvars.end())
-    {
-      new_subs.insert(new_subs.begin(), subvars.begin() + 1, subvars.end());
-    }
-    if (children_it != d_children.end())
+    Trace("sym-dt-debug") << "  merge indices ";
+    for (unsigned i : indices)
     {
-      (children_it->second).addNode(target_var, new_subs);
-    }
-    else
-    {
-      PartitionTrie pt;
-
-      pt.addNode(target_var, new_subs);
-      d_children[subvars[0]] = pt;
+      Trace("sym-dt-debug") << i << " ";
     }
+    Trace("sym-dt-debug") << std::endl;
   }
-}
+  Assert(!indices.empty());
 
-void SymmetryDetect::printPartition(Partition p)
-{
-  for (map<Node, vector<Node> >::iterator sub_vars_it =
-           p.d_subvar_to_vars.begin();
-       sub_vars_it != p.d_subvar_to_vars.end();
-       ++sub_vars_it)
-  {
-    Trace("sym-dt") << "[sym-dt] Partition: " << sub_vars_it->first << " -> {";
+  // initialize partition merger class
+  PartitionMerger pm;
+  pm.initialize(k, partitions, indices);
 
-    for (vector<Node>::iterator part_it = (sub_vars_it->second).begin();
-         part_it != (sub_vars_it->second).end();
-         ++part_it)
+  for (unsigned index : indices)
```

Comment:
Should we also categorize the "indices" here by types so that we only compare indices of the same type and reduce the amount of comparisons?

Reply:
Yes, good catch.  I think the code does not currently do this.  This indeed would be a good optimization. After merging this PR, I will add this.